### ISSUE TITLE
Regression bug? `briefcase build` installs packages in the wrong order in 0.3.3, but not 0.3.2

### ISSUE THREAD

**Issue by micahflee** [beeware/briefcase#504]
My project requires pyside2, which depends on libshiboken2.

With briefcase 0.3.3, when I run `briefcase create` and then `briefcase build`, it fails with this error:

```
$ briefcase build

[onionshare] Building AppImage...

[onionshare] Entering Docker context...
linuxdeploy version 1-alpha (git commit ID b6a1ba2), Travis build 488 built on 2020-10-09 13:07:48 UTC

-- Creating basic AppDir structure -- 
Creating directory /app/OnionShare/OnionShare.AppDir/usr/bin/ 
Creating directory /app/OnionShare/OnionShare.AppDir/usr/lib/ 

[...snip...]

Deploying files in directory /app/OnionShare/OnionShare.AppDir/usr/app_packages/PySide2 
Deploying dependencies for ELF file in AppDir: /app/OnionShare/OnionShare.AppDir/usr/app_packages/PySide2/QtOpenGL.abi3.so 
Deploying dependencies for ELF file /app/OnionShare/OnionShare.AppDir/usr/app_packages/PySide2/QtOpenGL.abi3.so 
ERROR: Could not find dependency: libshiboken2.abi3.so.5.15 
WARNING: Failed to deploy dependencies for ELF file /app/OnionShare/OnionShare.AppDir/usr/app_packages/PySide2/QtOpenGL.abi3.so, skipping 

Error while building app onionshare.
```

However I've confirmed that `libshiboken2.abi3.so.5.15` should in fact be there:

```
$ find linux/ |grep libshiboken2.abi3.so.5.15
linux/OnionShare/OnionShare.AppDir/usr/app_packages/shiboken2/libshiboken2.abi3.so.5.15
```

While debugging this I discovered the arguments passed into `linuxdeploy-x86_64.AppImage`. It looks like `--deploy-deps-only /app/OnionShare/OnionShare.AppDir/usr/app_packages/shiboken2` is near the end of the args, and `--deploy-deps-only /app/OnionShare/OnionShare.AppDir/usr/app_packages/PySide2` is near the beginning of the args.

Basically, I think the dependencies are being installed in the wrong order. `shiboken2` should be installed first so that when `PySide2` gets installed it can find the right libraries.

I wasn't sure if this was a briefcase bug or not, so I tried the same thing with briefcase 0.3.2 instead.

In version 0.3.2, it actually successfully builds my AppBundle! (However, it's not runnable because of #420 which was fixed in 0.3.3...)

I also just tried using briefcase from the master branch, but it still fails in the same way.

**Environment:**
 - Operating System: Ubuntu 20.04
 - Python version: 3.8.5
 - Software versions:
   - Briefcase: 0.3.3, 0.3.2

**Comment by micahflee** [comment]
I was just trying to see if I could fix this. It looks like this is where the bug happens: https://github.com/beeware/briefcase/blob/master/src/briefcase/platforms/linux/appimage.py#L184

But I don't really understand the purpose of this code. However I did reproduce it in a really simple way. Steps to reproduce:

- Start following the [Beeware Tutorial](https://docs.beeware.org/en/latest/tutorial/tutorial-0.html)
- When you run `briefcase new` and it asks for the GUI framework, choose PySide2
- Run `briefcase create`
- Run `briefcase build`, and it crashes the same way:

```
$ briefcase build

[helloworld] Building AppImage...

[helloworld] Entering Docker context...
linuxdeploy version 1-alpha (git commit ID b6a1ba2), Travis build 488 built on 2020-10-09 13:07:48 UTC

-- Creating basic AppDir structure -- 
Creating directory /app/Hello World/Hello World.AppDir/usr/bin/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/lib/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/share/applications/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/share/icons/hicolor/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/share/icons/hicolor/16x16/apps/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/share/icons/hicolor/32x32/apps/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/share/icons/hicolor/64x64/apps/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/share/icons/hicolor/128x128/apps/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/share/icons/hicolor/256x256/apps/ 
Creating directory /app/Hello World/Hello World.AppDir/usr/share/icons/hicolor/scalable/apps/ 

-- Deploying dependencies for existing files in AppDir -- 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/bin/python3.8 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libpthread.so.0 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libdl.so.2 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libutil.so.1 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libm.so.6 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libc.so.6 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_ssl.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /lib/x86_64-linux-gnu/libssl.so.1.0.0
Deploying copyright files for file /lib/x86_64-linux-gnu/libssl.so.1.0.0 
Deploying shared library /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
Deploying copyright files for file /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_gdbm.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /usr/lib/x86_64-linux-gnu/libgdbm.so.3
Deploying copyright files for file /usr/lib/x86_64-linux-gnu/libgdbm.so.3 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_lzma.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /lib/x86_64-linux-gnu/liblzma.so.5
Deploying copyright files for file /lib/x86_64-linux-gnu/liblzma.so.5 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_sqlite3.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /usr/lib/x86_64-linux-gnu/libsqlite3.so.0
Deploying copyright files for file /usr/lib/x86_64-linux-gnu/libsqlite3.so.0 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/unicodedata.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/zlib.cpython-38-x86_64-linux-gnu.so 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libz.so.1 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_codecs_tw.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_statistics.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/pyexpat.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_crypt.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /lib/x86_64-linux-gnu/libcrypt.so.1
Deploying copyright files for file /lib/x86_64-linux-gnu/libcrypt.so.1 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_sha512.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/syslog.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_posixshmem.cpython-38-x86_64-linux-gnu.so 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/librt.so.1 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_struct.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_socket.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_posixsubprocess.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_elementtree.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/select.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_csv.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_ctypes.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /usr/lib/x86_64-linux-gnu/libffi.so.6
Deploying copyright files for file /usr/lib/x86_64-linux-gnu/libffi.so.6 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/cmath.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/resource.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_codecs_jp.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/binascii.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_heapq.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_pickle.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_codecs_iso2022.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_dbm.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /usr/lib/x86_64-linux-gnu/libgdbm_compat.so.3
Deploying copyright files for file /usr/lib/x86_64-linux-gnu/libgdbm_compat.so.3 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/grp.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/nis.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /lib/x86_64-linux-gnu/libnsl.so.1
Deploying copyright files for file /lib/x86_64-linux-gnu/libnsl.so.1 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_multiprocessing.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_bz2.cpython-38-x86_64-linux-gnu.so 
Deploying shared library /lib/x86_64-linux-gnu/libbz2.so.1.0
Deploying copyright files for file /lib/x86_64-linux-gnu/libbz2.so.1.0 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_xxtestfuzz.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/audioop.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/mmap.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/math.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_opcode.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_codecs_hk.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_md5.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_codecs_kr.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_asyncio.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_codecs_cn.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/fcntl.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_blake2.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_datetime.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/ossaudiodev.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_sha3.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_sha1.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_xxsubinterpreters.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_json.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_hashlib.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_sha256.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_lsprof.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_queue.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/array.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_multibytecodec.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_random.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/parser.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/spwd.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_decimal.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_contextvars.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/termios.cpython-38-x86_64-linux-gnu.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/lib/python3.8/lib-dynload/_bisect.cpython-38-x86_64-linux-gnu.so 

-- Deploying dependencies only for ELF files -- 
Deploying files in directory /app/Hello World/Hello World.AppDir/usr/app_packages/PySide2 
Deploying dependencies for ELF file in AppDir: /app/Hello World/Hello World.AppDir/usr/app_packages/PySide2/QtOpenGL.abi3.so 
Deploying dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/app_packages/PySide2/QtOpenGL.abi3.so 
ERROR: Could not find dependency: libshiboken2.abi3.so.5.15 
WARNING: Failed to deploy dependencies for ELF file /app/Hello World/Hello World.AppDir/usr/app_packages/PySide2/QtOpenGL.abi3.so, skipping 

Error while building app helloworld.
```

**Comment by paulproteus** [comment]
Howdy @micahflee :D 

That's a great repro. I might have time to look into this next weekend (Oct 25/26), or perhaps a week or two after that. Curious what your timeline is, just to figure out if my random BeeWare volunteer time availability corresponds to your situation.

**Comment by jgirardet** [comment]
this has also been reported in #470 
Maybe this issue should be concerned by #485 ?

**Comment by micahflee** [comment]
Hi @paulproteus !! Thanks so much for offering to help work on this. I don't have a solid timeline, but basically if you can get it done before we are ready to release OnionShare 2.3, we can release an AppImage package. If not, we'll just release Snap and Flatpak packages.

**Comment by freakboy3742** [comment]
I'm going to close this issue to consolidate the general problem of "Linux AppImage deployment has problems" under a single issue (#718).
